docarray/array/mixins/match.py

```diff
@@ -65,170 +65,20 @@ def match(
-                      all Documents in `another` are considered as matches
-        :param normalization: a tuple [a, b] to be used with min-max normalization,
-                              the min distance will be rescaled to `a`, the max distance will be rescaled to `b`
-                              all values will be rescaled into range `[a, b]`.
-        :param batch_size: length of the chunks loaded into memory from darray.
-        :param metric_name: if provided, then match result will be marked with this string.
-        :param num_worker: the number of parallel workers. If not given, then the number of CPUs in the system will be used.
-        :return: distances and indices
-        """
-
-        x_mat = self.embeddings
-        n_x = x_mat.shape[0]
-
-        idx = 0
-        top_dists = np.inf * np.ones((n_x, limit))
-        top_inds = np.zeros((n_x, limit), dtype=int)
-
-        def _get_dist(da: 'DocumentArray'):
-            y_batch = da.embeddings
-
-            distances = cdist(x_mat, y_batch, metric_name)
-            dists, inds = top_k(distances, limit, descending=False)
-
-            if isinstance(normalization, (tuple, list)) and normalization is not None:
-                dists = minmax_normalize(dists, normalization)
-
-            return dists, inds, y_batch.shape[0]
-
-        if num_worker is None or num_worker > 1:
-            # notice that all most all computations (regardless the framework) are conducted in C
-            # hence there is no worry on Python GIL and the backend can be safely put to `thread` to
-            # save unnecessary data passing. This in fact gives a huge boost on the performance.
-            _gen = darray.map_batch(
-                _get_dist,
-                batch_size=batch_size,
-                backend='thread',
-                num_worker=num_worker,
-            )
-        else:
-            _gen = (_get_dist(b) for b in darray.batch(batch_size=batch_size))
-
-        for (dists, inds, _bs) in _gen:
-            inds += idx
-            idx += _bs
-            top_dists, top_inds = update_rows_x_mat_best(
-                top_dists, top_inds, dists, inds, limit
-            )
-
-        # sort final the final `top_dists` and `top_inds` per row
-        permutation = np.argsort(top_dists, axis=1)
-        dist = np.take_along_axis(top_dists, permutation, axis=1)
-        idx = np.take_along_axis(top_inds, permutation, axis=1)
-
-        return dist, idx
+        result = darray.search(
+            self,
+            metric=metric,
+            limit=limit,
+            normalization=normalization,
+            metric_name=metric_name,
+            batch_size=batch_size,
+            exclude_self=exclude_self,
+            only_id=only_id,
+            use_scipy=use_scipy,
+            device=device,
+            num_worker=num_worker,
+        )
+
+        for _r, _d in zip(result, self):
+            _d.matches = _r.matches
```

Comment:
why do we have to copy the query inside `search` and then fill the matches again at the level of match ?
I think search should return a list of match document arrays (if len(query) > 1) or a document array of matches (if len(query) == 1)

Reply:
> I also think that weaviate is not affected by this PR right ?

the weaviate `search` has been refactored in this PR